pkg/phlaredb/block_querier.go

```diff
@@ -905,6 +938,90 @@ func (b *singleBlockQuerier) SelectMatchingProfiles(ctx context.Context, params
 	return iter.NewMergeIterator(maxBlockProfile, false, iters...), nil
 }
 
+// Series selects the series labels from this block.
+//
+// Note: It will select ALL the labels in the block, not necessarily just the
+// subset in the time range SeriesRequest.Start to SeriesRequest.End.
+func (b *singleBlockQuerier) Series(ctx context.Context, params *ingestv1.SeriesRequest) ([]*typesv1.Labels, error) {
+	sp, ctx := opentracing.StartSpanFromContext(ctx, "Series Block")
+	defer sp.Finish()
+
+	err := b.Open(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	selectors, err := parseSelectors(params.Matchers)
+	if err != nil {
+		return nil, err
+	}
+
+	names, err := b.index.LabelNames()
+	if err != nil {
+		return nil, err
+	}
+
+	labelNamesFilter := make(map[string]struct{}, len(params.LabelNames))
+	for _, n := range params.LabelNames {
+		labelNamesFilter[n] = struct{}{}
+	}
+	names = lo.Filter(names, func(name string, _ int) bool {
+		_, ok := labelNamesFilter[name]
+		return ok
+	})
+
+	var postingsIters []iter.SeekIterator[storage.SeriesRef, storage.SeriesRef]
+	if selectors.matchesAll() {
+		for _, name := range names {
+			values, err := b.index.LabelValues(name)
+			if err != nil {
+				return nil, err
+			}
+
+			iter, err := b.index.Postings(name, nil, values...)
+			if err != nil {
+				return nil, err
+			}
+			postingsIters = append(postingsIters, iter)
+		}
+	} else {
+		for _, matchers := range selectors {
+			iter, err := PostingsForMatchers(b.index, nil, matchers...)
+			if err != nil {
+				return nil, err
+			}
+			postingsIters = append(postingsIters, iter)
+		}
+	}
```

Comment:
Depending on the selectors provided, collect posting iterators one of two ways:

1. If no selectors, use `b.index.Postings` for every label name in the request
2. If selectors, use `PostingsForMatchers` for each selector

This is done because `PostingsForMatchers` requires at least 1 selector. Sometimes none are provided. In that case, I collected an iterator for each label name.

Reply:
You might be fine using `PostingForMatchers` all the time see https://github.com/grafana/pyroscope/blob/15e9b770dd4a920431d9a74fd4af85a1f9e6894a/pkg/phlaredb/querier.go#L126-L127 which gets all posting in case no matcher is given

Reply:
That will only run if `notIts` has data. If you pass no matchers, then both `its` and `notIts` will be empty and that if block will never run.

Reply:
I am kind of surprised by this too. The reason why we never faced this issue, is because we do always use a `__profile_type__=` matcher for our SelectSeries queries.

I think we could potentially refactor the method to be like the prometheus where this was originally copied from  https://github.com/prometheus/prometheus/blob/0de7f39e6ad8ab43d1fac377e25f3b76cfaf616a/tsdb/querier.go#L240 then it would support it and also come with other optimizations.